infrastructure/src/main/java/org/corfudb/infrastructure/log/Segment.java

```diff
@@ -0,0 +1,557 @@
+     */
+    private LogFormat.LogHeader parseHeader(FileChannel channel) throws IOException {
+        LogFormat.Metadata metadata = parseMetadata(channel);
+        if (metadata == null) {
+            // Partial write on the metadata for the header
+            // Rewind the channel position to the beginning of the file
+            channel.position(0);
+            return null;
+        }
+
+        ByteBuffer buffer = getPayloadForMetadata(channel, metadata);
+        if (buffer == null) {
+            // partial write on the header payload
+            // Rewind the channel position to the beginning of the file
+            channel.position(0);
+            return null;
+        }
+
+        if (Checksum.getChecksum(buffer.array()) != metadata.getPayloadChecksum()) {
+            String errorMessage = getDataCorruptionErrorMessage("Invalid metadata checksum",
+                    channel, segmentFilePath
+            );
+            throw new DataCorruptionException(errorMessage);
+        }
+
+        LogFormat.LogHeader header;
+
+        try {
+            header = LogFormat.LogHeader.parseFrom(buffer.array());
+        } catch (InvalidProtocolBufferException e) {
+            String errorMessage = getDataCorruptionErrorMessage("Invalid header",
+                    channel, segmentFilePath
+            );
+            throw new DataCorruptionException(errorMessage, e);
+        }
+
+        return header;
+    }
+
+    public synchronized void retain() {
+        refCount++;
+    }
+
+    public synchronized void release() {
+        if (refCount == 0) {
+            throw new IllegalStateException("refCount cannot be less than 0, segment " + id);
+        }
+        refCount--;
+    }
+
+    public void close() {
+
+        Set<FileChannel> channels = new HashSet<>(
+                Arrays.asList(writeChannel, readChannel)
+        );
+
+        for (FileChannel channel : channels) {
+            try {
+                channel.force(true);
+            } catch (IOException e) {
+                log.error("Can't force updates for {}", segmentFilePath, e);
+            } finally {
+                IOUtils.closeQuietly(channel);
+            }
+        }
+
+        if (refCount != 0) {
+            log.warn("closeSegmentHandlers: Segment {} is trimmed, but refCount is {}, attempting to trim anyways",
+                    segmentFilePath, refCount);
+        }
+    }
+
+    /**
+     * This index maps a sequence number to a file offset + payload length.
+     *
+     * The file offset and payload length are packed in a long (treated as unsigned) as follows:
+     *
+     *  |--------64-bits--------|
+     *  |fileOffset|payloadSize|
+     *
+     */
+    private static final class Index {
+
+        private final BoundedMap map;
+        private final int highBitsNum;
+
+        private Index(long offset, int size) {
+            this.map = new BoundedMap(offset, size);
+            this.highBitsNum = Long.bitCount(MAX_SEGMENT_SIZE);
```

Comment:
I think the packing and unpacking here is not clear.
`highBitsNum` here is 36, which tells that the highest 36 bits is for file offset and the lowest 28 bits is for payload size.
But `(((long) high) << this.highBitsNum) | (low & MAX_WRITE_SIZE);` shows that the highest 28 bits are for file offset and the lowest 36 bits are for payload size. Actually after masking the low is taking the lowest 28 bits.

Reply:
`(((long) high) << this.highBitsNum)` will shift left by 36, leaving the lower 28 bits zeroed out, then `(low & MAX_WRITE_SIZE)` will keep the lower 28 bits, and ORing them will be:

```
       |--------64-bits--------|
       |fileOffset|payloadSize|
```

Reply:
Wait, are you saying its incorrect, or are you saying that the confusing part is that the applied mask is in hex, but the shift is in bits?

Reply:
Yes, you are correct.

I incorrectly copied the correct code from here (same patch):
https://github.com/CorfuDB/CorfuDB/pull/3533/files#diff-bb37cc7aa935dcd8db14f794b584a15047ca900c76cf534a6747f6bc16400e3bR31

Which is also tested here:
https://github.com/CorfuDB/CorfuDB/pull/3533/files#diff-13a26acd7a2fa2a902f2b460858dfce37ec05357b80cff13f8699d80645f69ebR25


Will migrate the tests and fix this.